components/raftstore/src/store/peer_storage.rs

```diff
@@ -396,6 +396,51 @@ pub fn init_apply_state(engines: &Engines, region: &Region) -> Result<RaftApplyS
     )
 }
 
+fn validate_states(region_id: u64, engines: &Engines, raft_state: &mut RaftLocalState, apply_state: &RaftApplyState) -> Result<()> {
+    let last_index = raft_state.get_last_index();
+    let mut commit_index = raft_state.get_hard_state().get_commit();
+    let apply_index = apply_state.get_applied_index();
+
+    if commit_index < apply_state.get_last_commit_index() {
+        return Err(box_err!(
+            "raft state {:?} not match apply state {:?} and can't be recovered.",
+            raft_state,
+            apply_state
+        ));
+    }
+    let recorded_commit_index = apply_state.get_commit_index();
+    if commit_index < recorded_commit_index {
+        let log_key = keys::raft_log_key(region_id, recorded_commit_index);
+        let entry = engines.raft.get_msg::<Entry>(&log_key)?;
+        if entry.map_or(true, |e| e.get_term() != apply_state.get_commit_term()) {
```

Comment:
May this entry have been deleted by raft_log_gc?

Reply:
It can't. If it's deleted, it means the log should be committed. Hence commit index should be larger than recorded_commit_index if there is no data loss.